src/basic/fileio.c

```diff
@@ -259,7 +259,14 @@ int read_full_stream(FILE *f, char **contents, size_t *size) {
                 if (ferror(f))
                         return -errno;
 
-                if (feof(f))
+                /* stdio will only set EOF after attempting to read past the
+                 * end of the file.  In the common regular file case where
+                 * know, allocate, and read the file's exact size, EOF isn't
+                 * found yet.  Do our own check against size here to avoid
+                 * doubling the buffer just for fread() to read nothing and
+                 * find eof on the next iteration.
+                 */
+                if (feof(f) || (st_size > 0 && st_size == l))
```

Comment:
I am not sure I like this change. In the original code I tried to never mix the size data we gather from stat() and the size we actually read. We use st_size as optimization for sizing the initial buffer, and we use it as an early-exit condition, but we do not use it in any relevant control flow for deciding when to stop reading. This basically means that it's either the actual data we read that counts, or the file size reported through stat(), but we never intermix the two. The control flow paths through the function either check one OR the other, (though possibly use one as an optimized start-value for the other, but never as something that actually mattered), but never both in conjunction.

I think this is a good approach for two reasons: in file systems such as /proc or /sys the reported file size generally doesn't match what you actually can read from a file (of course you could argue that this special case will always report exactly st_size == 0, but still...), and secondly, it has a bit of "racy" feel to it. Of course, reading the file iteratively in blocks is racy anyway and files can't be read atomically this way, but at least the order in which things are done was previously clearly defined: we'd start from the beginning of the file, and read sequentially until it is reported that we reached the end. The EOF condition was thus a result of the last read, and hence neatly ordered against the previous read. But if we would now use the stat() data for detecting EOF, then we'd use information that was acquired before we started the loop much much later, after quite some time has passed, after quite a number of iterations have completed. Hence the exit condition with this change would not match what was actually read anymore.

Of course, neither of this is really the strongest reason in the universe, but it still makes me feel uneasy about this patch.

I think a different approach to optimize the reading here might be a better option: in the first iteration, let's try to read 1 byte more than the size reported by stat(). Of course, if the file didn't get changed since the stat(), then this should result in a short reach (after all, fread() only results in short reads in case of EOF), and feof() should reported true immediately after we come back from fread(). The net effect is pretty much the same: we'd iterate at most once through the loop, unless something's strange with the file.

I hope this makes sense?